### PR TITLE
Postprocessing

### PR DESCRIPTION
**Pull request by numberZero**
The post-processing shader is very simple currently, but that’s only the beginning. And even that simple shader may improve performance greatly sometimes:

**Standard tone mapping:**
plain

**Tone mapping at post-processing:**
postprocessing

The reason here seems to be that there are TONS of `plantlike` leaves, so usual shader gets called way more often than necessary. The performance difference is smaller when undersampling is in effect, however. Also, there is a side effect: sky color is affected too now. But that’s just an optin anyway.

### CODE REVIEW

File: src/drawscene.cpp [hunk 1]
```diff
@@ -498,12 +506,45 @@ void draw_plain(Camera &camera, bool show_hud,
...
+	// Post-process
+	if (undersampling || postprocessing) {
+		driver->setRenderTarget(0, false, false);
+		static video::S3DVertex vertices[4] = {
+			video::S3DVertex(1.0, -1.0, 0.0, 0.0, 0.0, -1.0, video::SColor(255, 0, 255, 255), 1.0, 0.0),
+			video::S3DVertex(-1.0, -1.0, 0.0, 0.0, 0.0, -1.0, video::SColor(255, 255, 0, 255), 0.0, 0.0),
+			video::S3DVertex(-1.0, 1.0, 0.0, 0.0, 0.0, -1.0, video::SColor(255, 255, 255, 0), 0.0, 1.0),
+			video::S3DVertex(1.0, 1.0, 0.0, 0.0, 0.0, -1.0, video::SColor(255, 255, 255, 255), 1.0, 1.0),
+		};
+		static u16 indices[6] = { 0, 1, 2, 2, 3, 0 };
```
Comment (nerzhul) [inline src/drawscene.cpp]: missing constness

File: src/drawscene.cpp [hunk 2]
```diff
@@ -498,12 +506,45 @@ void draw_plain(Camera &camera, bool show_hud,
...
-	// Upscale lowres render
-	if (undersampling > 0) {
-		driver->setRenderTarget(0, true, true);
-		driver->draw2DImage(image,
-				irr::core::rect<s32>(0, 0, dest_size.X, dest_size.Y),
-				irr::core::rect<s32>(0, 0, pixelated_size.X, pixelated_size.Y));
+	// Post-process
+	if (undersampling || postprocessing) {
+		driver->setRenderTarget(0, false, false);
+		static video::S3DVertex vertices[4] = {
```
Comment (nerzhul) [inline src/drawscene.cpp]: constness

File: src/drawscene.cpp (hunk 1 above)
Comment (numberZero) [inline src/drawscene.cpp]: fixed

**Comment by numberZero** [comment]
This part seems to be finished. Next parts on my queue:
- Depth of field
- Gamma correction
- Refactoring
- Applying effects in 3D modes

but to ease reviewing and prevent stalling, I won’t place everything in a single PR. (I can’t, actually, as there may be dependencies)

**Comment by paramat** [comment]
Note that RBA's DoF was completely wrong, it had unrealistic proportions of depth of focus in front and behind the focus. Study of optics is needed.

**Comment by numberZero** [comment]
@paramat his DoF was very strange, really. My version is completely different, although may be slower.
@nerzhul I fixed what you mentioned.

**Comment by rubenwardy** [comment]
I'm very interested in an PRs which will help bring us closer to this https://youtu.be/Rkydse1LVX0
Although it shouldn't come at the cost of performance for all users. Between fancy effects and view range I'd go for view range

**Comment by Fixer-007** [comment]
Shaders both in minetest and minecraft bring performance down. Fancy shaders I've tried in minecraft hit fps especially hard down to 20-30 fps levels on AMD HD6870, same for that RBA's branch with bloom, artistic DOF, fps plunged to 30.
I would rather see proper z-sorting (blinking is annoying) and improving general rendering speed (minetest bedrock engine with 60fps on vrange 1024, unthinkable for minetest)

**Comment by rubenwardy** [comment]
> I would rather see proper z-sorting (blinking is annoying) and improving general rendering speed (minecraft bedrock engine with 60fps on vrange 1024, unthinkable for minetest)

I agree. You say it's unthinkable, but I think we should aim for it

**Comment by numberZero** [comment]
This PR is obsolete since https://github.com/minetest/minetest/commit/28841961ba91b943b7478704181604fa3e24e81e.

**Comment by appgurueu** [comment]
Post-processing is very mighty. Maybe we should render to a second FBO while rendering the scene, creating a depthmap out of the players view, what makes us able to implement an edgefinder or focus/blur etc..., as here : https://forum.minetest.net/viewtopic.php?f=5&t=18005&start=25#p301838
Also, some parts of the post-processing shader of my "engine" may be useful, such as blur or edgefinder : https://github.com/appgurueu/new-shaders/blob/master/2d_fs.glsl

**Comment by appgurueu** [comment]
> 
> 
>     I would rather see proper z-sorting (blinking is annoying) and improving general rendering speed (minecraft bedrock engine with 60fps on vrange 1024, unthinkable for minetest)
> 
> I agree. You say it's unthinkable, but I think we should aim for it

Maybe that could be done in a geometry shader ?

**Pull request closed without merging**